Blocky Crashing with SEGV

### Discussed in https://github.com/0xERR0R/blocky/discussions/776

<div type='discussions-op-text'>

<sup>Originally posted by **kbrierly** December  1, 2022</sup>
Is anyone else having issues with blocky in docker crashing with a SEGV?
this is with the 0.20 version using the docker hub container,
my upstream is DOT to 9.9.9.9,
i have custom domains and conditional forwarding enabled 


```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xcbbe38]

goroutine 2728 [running]:
github.com/0xERR0R/blocky/resolver.addExtraReasoning(0xc0000365a0?)
        /go/src/resolver/ede_resolver.go:42 +0x18
github.com/0xERR0R/blocky/resolver.(*EdeResolver).Resolve(0xc00107f920, 0xc000014b48?)
        /go/src/resolver/ede_resolver.go:24 +0x48
github.com/0xERR0R/blocky/resolver.(*ClientNamesResolver).Resolve(0xc0000362d0, 0xc00143c5b0)
        /go/src/resolver/client_names_resolver.go:79 +0x18c
github.com/0xERR0R/blocky/resolver.(*FqdnOnlyResolver).Resolve(0x4?, 0x4?)
        /go/src/resolver/fqdn_only_resolver.go:34 +0xf4
github.com/0xERR0R/blocky/resolver.(*FilteringResolver).Resolve(0xc00107f8f0, 0xc00143c5b0)
        /go/src/resolver/filtering_resolver.go:29 +0x1bc
github.com/0xERR0R/blocky/server.(*Server).OnRequest(0xc0002ac0f0, {0x108b978, 0xc000aa5b00}, 0xc001428ab0)
        /go/src/server/server.go:617 +0xee
github.com/miekg/dns.HandlerFunc.ServeDNS(0xc00012d500?, {0x108b978?, 0xc000aa5b00?}, 0xc00142001c?)
        /go/pkg/mod/github.com/miekg/dns@v1.1.50/server.go:37 +0x2f
github.com/miekg/dns.(*ServeMux).ServeDNS(0xc0012ca000?, {0x108b978?, 0xc000aa5b00?}, 0xc001428ab0?)
        /go/pkg/mod/github.com/miekg/dns@v1.1.50/serve_mux.go:103 +0x72
github.com/miekg/dns.(*Server).serveDNS(0xc0001785a0, {0xc0012ca000, 0x34, 0xffff}, 0xc000aa5b00)
        /go/pkg/mod/github.com/miekg/dns@v1.1.50/server.go:659 +0x462
github.com/miekg/dns.(*Server).serveUDPPacket(0xc0001785a0, 0x0?, {0xc0012ca000, 0x34, 0xffff}, {0x10897f0?, 0xc000504008}, 0xc0013ad4c0, {0x0, 0x0})
        /go/pkg/mod/github.com/miekg/dns@v1.1.50/server.go:603 +0x1dc
created by github.com/miekg/dns.(*Server).serveUDP
        /go/pkg/mod/github.com/miekg/dns@v1.1.50/server.go:533 +0x485
```
</div>

--- a/resolver/ede_resolver.go
+++ b/resolver/ede_resolver.go
@@ -19,6 +19,9 @@ func NewEdeResolver(cfg config.EdeConfig) ChainedResolver {
 
 func (r *EdeResolver) Resolve(request *model.Request) (*model.Response, error) {
 	resp, err := r.next.Resolve(request)
+	if err != nil {
+		return nil, err
+	}
 
 	if r.config.Enable {
 		addExtraReasoning(resp)

--- a/resolver/ede_resolver_test.go
+++ b/resolver/ede_resolver_test.go
@@ -1,6 +1,8 @@
 package resolver
 
 import (
+	"errors"
+
 	"github.com/0xERR0R/blocky/config"
 	"github.com/0xERR0R/blocky/model"
 
@@ -23,12 +25,14 @@ var _ = Describe("EdeResolver", func() {
 	})
 
 	JustBeforeEach(func() {
-		m = &MockResolver{}
-		m.On("Resolve", mock.Anything).Return(&model.Response{
-			Res:    mockAnswer,
-			RType:  model.ResponseTypeCUSTOMDNS,
-			Reason: "Test",
-		}, nil)
+		if m == nil {
+			m = &MockResolver{}
+			m.On("Resolve", mock.Anything).Return(&model.Response{
+				Res:    mockAnswer,
+				RType:  model.ResponseTypeCUSTOMDNS,
+				Reason: "Test",
+			}, nil)
+		}
 
 		sut = NewEdeResolver(sutConfig).(*EdeResolver)
 		sut.Next(m)
@@ -82,6 +86,21 @@ var _ = Describe("EdeResolver", func() {
 			c := sut.Configuration()
 			Expect(c).Should(HaveLen(1))
 			Expect(c[0]).Should(Equal("activated"))
+		})
+
+		When("resolver returns an error", func() {
+			resolveErr := errors.New("test")
+
+			BeforeEach(func() {
+				m = &MockResolver{}
+				m.On("Resolve", mock.Anything).Return(nil, resolveErr)
+			})
+
+			It("should return it", func() {
+				resp, err := sut.Resolve(newRequest("example.com", dns.Type(dns.TypeA)))
+				Expect(resp).To(BeNil())
+				Expect(err).To(Equal(resolveErr))
+			})
 		})
 	})
 })

